File: firmware/egble-controller/Protocol.cpp
#include "Protocol.h"
#include <Arduino.h>
#include <ArduinoJson.h>
#include <string.h>
#include <strings.h>

// Fill a PatternConfig from a command object. Only fields present in the JSON
// override the defaults, so a caller can send just the ones they care about.
static PatternConfig configFromJson(JsonObjectConst o) {
  PatternConfig c;
  const char* typeName = o["type"] | "";
  PatternType t = patternFromName(typeName);
  if (t != PAT_COUNT) c.type = t;

  if (o["bri"].is<int>())       c.bri       = (uint8_t)constrain(o["bri"].as<int>(), 0, 255);
  if (o["onMs"].is<int>())      c.onMs      = (uint16_t)o["onMs"].as<int>();
  if (o["offMs"].is<int>())     c.offMs     = (uint16_t)o["offMs"].as<int>();
  if (o["fadeInMs"].is<int>())  c.fadeInMs  = (uint16_t)o["fadeInMs"].as<int>();
  if (o["holdMs"].is<int>())    c.holdMs    = (uint16_t)o["holdMs"].as<int>();
  if (o["fadeOutMs"].is<int>()) c.fadeOutMs = (uint16_t)o["fadeOutMs"].as<int>();
  if (o["gapMs"].is<int>())     c.gapMs     = (uint16_t)o["gapMs"].as<int>();
  if (o["stepMs"].is<int>())    c.stepMs    = (uint16_t)o["stepMs"].as<int>();
  if (o["overlapMs"].is<int>()) c.overlapMs = (uint16_t)o["overlapMs"].as<int>();
  return c;
}

static ProtocolResult ack(bool changed, bool ok, const char* msg) {
  ProtocolResult r;
  r.changed = changed;
  r.ok = ok;
  strncpy(r.msg, msg, sizeof(r.msg) - 1);
  return r;
}

ProtocolResult handleCommand(const char* json, PatternEngine& eng, SceneManager& scenes) {
  JsonDocument doc;
  DeserializationError err = deserializeJson(doc, json);
  if (err) return ack(false, false, "bad json");

  const char* cmd = doc["cmd"] | "";

  if (strcmp(cmd, "pattern") == 0) {
    PatternConfig c = configFromJson(doc.as<JsonObjectConst>());
    if (doc["all"].is<bool>() && doc["all"].as<bool>()) {
      eng.setPatternAll(c);
      return ack(true, true, "pattern all");
    }
    if (doc["grp"].is<int>()) {
      eng.setPatternGroup((uint8_t)doc["grp"].as<int>(), c);
      return ack(true, true, "pattern group");
    }
    if (doc["ch"].is<int>()) {
      int ch = doc["ch"].as<int>();
      if (ch < 0 || ch >= eng.count()) return ack(false, false, "bad ch");
      eng.setPattern((uint8_t)ch, c);
      return ack(true, true, "pattern ch");
    }
    return ack(false, false, "no target");
  }

  if (strcmp(cmd, "enable") == 0) {
    int ch = doc["ch"] | -1;
    if (ch < 0 || ch >= eng.count()) return ack(false, false, "bad ch");
    eng.setEnabled((uint8_t)ch, doc["on"] | true);
    return ack(true, true, "enable");
  }

  if (strcmp(cmd, "group") == 0) {
    int ch = doc["ch"] | -1;
    if (ch < 0 || ch >= eng.count()) return ack(false, false, "bad ch");
    eng.setGroup((uint8_t)ch, (uint8_t)(doc["gid"] | 0));
    return ack(true, true, "group");
  }

  if (strcmp(cmd, "scale") == 0) {
    int ch = doc["ch"] | -1;
    if (ch < 0 || ch >= eng.count()) return ack(false, false, "bad ch");
    eng.setScale((uint8_t)ch, (uint8_t)constrain((int)(doc["pct"] | 100), 0, 100));
    return ack(true, true, "scale");
  }

  if (strcmp(cmd, "scene") == 0) {
    const char* action = doc["action"] | "";
    if (strcmp(action, "save") == 0) {
      int slot = doc["slot"] | -1;
      if (slot < 0 || slot >= MAX_SCENE_SLOTS) return ack(false, false, "bad slot");
      Scene s = scenes.capture(eng, doc["name"] | "Scene");
      bool ok = scenes.saveSlot((uint8_t)slot, s);
      return ack(false, ok, ok ? "scene saved" : "save failed");
    }
    if (strcmp(action, "load") == 0) {
      int slot = doc["slot"] | -1;
      Scene s;
      if (slot < 0 || slot >= MAX_SCENE_SLOTS || !scenes.loadSlot((uint8_t)slot, s))
        return ack(false, false, "load failed");
      scenes.apply(eng, s);
      return ack(true, true, "scene loaded");
    }
    if (strcmp(action, "delete") == 0) {
      int slot = doc["slot"] | -1;
      if (slot < 0 || slot >= MAX_SCENE_SLOTS) return ack(false, false, "bad slot");
      scenes.deleteSlot((uint8_t)slot);
      return ack(false, true, "scene deleted");
    }
    if (strcmp(action, "recall") == 0) {
      Scene s;
      if (!SceneManager::builtin(doc["name"] | "", s)) return ack(false, false, "no builtin");
      scenes.apply(eng, s);
      return ack(true, true, "builtin recalled");
    }
    return ack(false, false, "bad action");
  }

  if (strcmp(cmd, "trigger") == 0) {
    const char* action = doc["action"] | "";
    PatternConfig c;
    if (strcmp(action, "sos") == 0) {
      c.type = PAT_SOS; c.bri = 255; c.onMs = 200;
      eng.setPatternAll(c);
      return ack(true, true, "trigger sos");
    }
    if (strcmp(action, "left") == 0 || strcmp(action, "right") == 0) {
      c.type = PAT_TURN_SIGNAL; c.bri = 255; c.onMs = 180; c.offMs = 180;
      eng.setPatternAll(c);
      return ack(true, true, "trigger turn");
    }
    if (strcmp(action, "stop") == 0 || strcmp(action, "off") == 0) {
      c.type = PAT_OFF;
      eng.setPatternAll(c);
      return ack(true, true, "trigger stop");
    }
    return ack(false, false, "bad action");
  }

  if (strcmp(cmd, "get") == 0) {
    return ack(true, true, "get");  // caller pushes state
  }

  return ack(false, false, "unknown cmd");
}

size_t buildState(char* out, size_t outLen, const PatternEngine& eng) {
  JsonDocument doc;
  doc["v"] = FW_VERSION;
  JsonArray arr = doc["ch"].to<JsonArray>();
  for (uint8_t i = 0; i < eng.count(); i++) {
    const ChannelRuntime& rt = eng.runtime(i);
    JsonObject o = arr.add<JsonObject>();
    o["e"]   = rt.enabled ? 1 : 0;
    o["g"]   = rt.groupId;
    o["p"]   = patternName(rt.cfg.type);
    o["bri"] = rt.cfg.bri;
    o["sc"]  = eng.scale(i);
    o["lvl"] = eng.currentLevel(i);
  }
  return serializeJson(doc, out, outLen);
}

size_t buildScenes(char* out, size_t outLen, const SceneManager& scenes) {
  JsonDocument doc;
  JsonArray builtins = doc["builtin"].to<JsonArray>();
  for (uint8_t i = 0; i < SceneManager::builtinCount(); i++) {
    builtins.add(SceneManager::builtinName(i));
  }
  JsonArray slots = doc["slots"].to<JsonArray>();
  for (uint8_t i = 0; i < MAX_SCENE_SLOTS; i++) {
    JsonObject o = slots.add<JsonObject>();
    o["slot"] = i;
    if (scenes.slotUsed(i)) {
      char name[24];
      scenes.slotName(i, name, sizeof(name));
      o["name"] = name;
    } else {
      o["name"] = nullptr;
    }
  }
  return serializeJson(doc, out, outLen);
}

File: app/src/ble/protocol.ts
/**
 * EGBLE - command builders and state parsing
 *
 * Pure functions with no BLE or React dependency, so the wire format can be
 * unit-tested in isolation. Every builder returns a JSON string ready to write
 * to the command (or scene) characteristic. Shapes match docs/protocol.md.
 */

import { PatternConfig, PatternType } from '../constants/patterns';

// ---- Commands ----

/** Build a pattern command for a single channel. */
export function cmdPatternChannel(ch: number, cfg: PatternConfig): string {
  return JSON.stringify({ cmd: 'pattern', ch, ...compactPattern(cfg) });
}

/** Build a pattern command for a whole group. */
export function cmdPatternGroup(gid: number, cfg: PatternConfig): string {
  return JSON.stringify({ cmd: 'pattern', grp: gid, ...compactPattern(cfg) });
}

/** Build a pattern command targeting all channels. */
export function cmdPatternAll(cfg: PatternConfig): string {
  return JSON.stringify({ cmd: 'pattern', all: true, ...compactPattern(cfg) });
}

export function cmdEnable(ch: number, on: boolean): string {
  return JSON.stringify({ cmd: 'enable', ch, on });
}

export function cmdGroup(ch: number, gid: number): string {
  return JSON.stringify({ cmd: 'group', ch, gid });
}

export function cmdScale(ch: number, pct: number): string {
  return JSON.stringify({ cmd: 'scale', ch, pct: clamp(pct, 0, 100) });
}

export type TriggerAction = 'left' | 'right' | 'sos' | 'stop';

export function cmdTrigger(action: TriggerAction): string {
  return JSON.stringify({ cmd: 'trigger', action });
}

export function cmdGet(): string {
  return JSON.stringify({ cmd: 'get' });
}

// ---- Scene commands ----

export function cmdSceneSave(slot: number, name: string): string {
  return JSON.stringify({ cmd: 'scene', action: 'save', slot, name });
}

export function cmdSceneLoad(slot: number): string {
  return JSON.stringify({ cmd: 'scene', action: 'load', slot });
}

export function cmdSceneDelete(slot: number): string {
  return JSON.stringify({ cmd: 'scene', action: 'delete', slot });
}

export function cmdSceneRecall(name: string): string {
  return JSON.stringify({ cmd: 'scene', action: 'recall', name });
}

// Only send the fields the given pattern type needs, plus its type. This keeps
// packets small and avoids overwriting firmware defaults the app did not set.
function compactPattern(cfg: PatternConfig): Record<string, unknown> {
  const out: Record<string, unknown> = { type: cfg.type };
  const include = (k: keyof PatternConfig) => {
    out[k] = cfg[k];
  };
  switch (cfg.type) {
    case 'SOLID':
      include('bri');
      break;
    case 'BLINK':
    case 'TURN_SIGNAL':
      include('bri');
      include('onMs');
      include('offMs');
      break;
    case 'FADE_PULSE':
      include('bri');
      include('fadeInMs');
      include('holdMs');
      include('fadeOutMs');
      include('gapMs');
      break;
    case 'SOS':
      include('bri');
      include('onMs');
      break;
    case 'SEQUENCE':
      include('bri');
      include('stepMs');
      include('overlapMs');
      break;
    case 'OFF':
    default:
      break;
  }
  return out;
}

// ---- State parsing ----

export interface ChannelState {
  enabled: boolean;
  groupId: number;
  pattern: PatternType;
  bri: number;
  scale: number;
  level: number;
}

/** Battery of the mounted swap cell, as reported by the inverter's VBAT sense. */
export interface BatteryState {
  present: boolean;   // did the device report a battery reading at all
  pct: number;        // 0..100 state of charge
  charging: boolean;  // on the charge pad / pigtail
}

export const ABSENT_BATTERY: BatteryState = { present: false, pct: 0, charging: false };

export type BatteryTier = 'good' | 'low' | 'critical';

/** Map a state-of-charge to the white / orange / red tier used everywhere. */
export function batteryTier(pct: number): BatteryTier {
  if (pct < 15) return 'critical';
  if (pct < 40) return 'low';
  return 'good';
}

export interface DeviceState {
  version: string;
  channels: ChannelState[];
  battery: BatteryState;
}

/** Parse a state notification payload. Returns null if it does not parse. */
export function parseState(json: string): DeviceState | null {
  try {
    const o = JSON.parse(json);
    if (!o || !Array.isArray(o.ch)) return null;
    // Battery is optional: older firmware / boards without VBAT sense omit `bat`.
    const bat = o.bat && typeof o.bat === 'object'
      ? { present: true, pct: clamp(Number(o.bat.p) || 0, 0, 100), charging: o.bat.ch === 1 || o.bat.ch === true }
      : ABSENT_BATTERY;
    return {
      version: typeof o.v === 'string' ? o.v : '',
      battery: bat,
      channels: o.ch.map((c: any) => ({
        enabled: c.e === 1 || c.e === true,
        groupId: Number(c.g) || 0,
        pattern: (c.p as PatternType) ?? 'OFF',
        bri: Number(c.bri) || 0,
        scale: typeof c.sc === 'number' ? c.sc : 100,
        level: Number(c.lvl) || 0,
      })),
    };
  } catch {
    return null;
  }
}

export interface SceneList {
  builtins: string[];
  slots: { slot: number; name: string | null }[];
}

/** Parse the scene characteristic read payload. */
export function parseScenes(json: string): SceneList | null {
  try {
    const o = JSON.parse(json);
    return {
      builtins: Array.isArray(o.builtin) ? o.builtin : [],
      slots: Array.isArray(o.slots)
        ? o.slots.map((s: any) => ({ slot: Number(s.slot), name: s.name ?? null }))
        : [],
    };
  } catch {
    return null;
  }
}

function clamp(v: number, min: number, max: number): number {
  return Math.max(min, Math.min(max, Math.round(v)));
}
